### ISSUE TITLE
batchnorm with BF16 implementation 

### ISSUE THREAD

**Issue by rabeehk** [deepmind/dm-haiku#396]
Hi
I implemented BF16.activations mode, this works fine, but when I add batchnorm to the network, this results in NaN values. I am not using jm.policy (I implemented down-cast/up-casts of params/grads ....  in a careful implementation) and I am running on TPUs. What should I do to handle this case?

 thanks.

Best regards,
Rabeeh

**Comment by tomhennigan** [comment]
Hi @rabeehk, is sounds like you're doing the right thing if you want bf16 activations for every module.

bf16 activations should work with BatchNorm, the Haiku imagenet example used to do precisely what you are describing (see: deepmind/dm-haiku@1e2b68d8) and trained just fine.

I suspect the issue here is specific to your network, data or training setup. I would suggest debugging this using JAX's `checkify` utility which will allow you to see precisely where in your network the NaN is being produced. The author of checkify @LenaMartens is at DeepMind, you may want to ask her about the utility on internal Slack during the working week.

If you want a workaround then I'd suggest upcasting to f32 around the batch norm module (and avoid casting batch norm params and state).

```python
def selective_bf16(x):
  return  hk.data_structures.map(lambda m, n, v: v.astype(jnp.bfloat16) if 'batch_norm' not in m else v, x)

def train_step(params, state, ...):
  # We want bf16 activations everywhere but BatchNorm.
  params = selective_bf16(params)
  state = selective_bf16(state)
  logits = f.apply(params, state, ...)
  ...

def f(x):
  assert x.dtype == jnp.bfloat16
  x = SomeModule()(x)

  # Do BatchNorm in f32.
  x = x.astype(jnp.float32)
  x = hk.BatchNorm(..)(x, ..)
  x = x.astype(jnp.bfloat16)

  ...
```

If it is a pain to find all the BatchNorm instances in your code you can use an interceptor to do this automatically:

```python
def bn_interceptor(next_f, args, kwargs, context):
  """Ensure BatchNorm is done in f32."""
  is_bn =  type(context.module) is hk.BatchNorm
  if is_bn:
    x = args[0]
    in_dtype = x.dtype
    x = x.astype(jnp.float32)
    args = (x, args[1:])
  y = next_f(*args, **kwargs)
  if is_bn:
    y = y.astype(in_dtype)
  return y

def f(x):
  with hk.intercept_methods(bn_interceptor):
    x = SomeModule()(x)
    x = hk.BatchNorm(..)(x, ..)
    ...

```

If you end up needing this kind of "selective casting" then I would consider using [`jmp`](https://github.com/deepmind/jmp). This (and loss scaling + skipping non-finite updates which you would need if using f16) are the main reasons why that library exists.

**Comment by rabeehk** [comment]
Hi Tom
thank you. The easier way working for me to upcast values coming to batchnorm and down-cast the final output of batchnorm again.
Best
Rabeeh
